File: python3/bron_kerbosch3.py
# coding: utf-8

from bron_kerbosch_pivot import pick_arbitrary, visit
from graph import UndirectedGraph, Vertex
from graph_degeneracy import degeneracy_ordering
from reporter import Reporter
from typing import Set


def bron_kerbosch3(graph: UndirectedGraph, reporter: Reporter) -> None:
    '''Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
       choosing a pivot arbitrarily'''
    excluded: Set[Vertex] = set()
    for v in degeneracy_ordering(graph=graph, drop=1):
        neighbours = graph.adjacencies[v]
        assert neighbours
        if neighbouring_candidates := neighbours.difference(excluded):
            neighbouring_excluded = neighbours.intersection(excluded)
            visit(graph=graph,
                  reporter=reporter,
                  initial_pivot_choice=pick_arbitrary,
                  further_pivot_choice=pick_arbitrary,
                  candidates=neighbouring_candidates,
                  excluded=neighbouring_excluded,
                  clique=[v])
        else:
            assert not excluded.isdisjoint(neighbours)
        excluded.add(v)

File: scala/src/main/scala/bron_kerbosch3.scala
// Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
// choosing a pivot arbitrarily

import bron_kerbosch_pivot.PivotChoice.Arbitrary
import bron_kerbosch_pivot.visit

import scala.collection.immutable

class bron_kerbosch3[Vertex: Integral] extends bron_kerbosch_algorithm[Vertex] {
  def explore(
      graph: UndirectedGraph[Vertex],
      reporter: immutable.Iterable[Vertex] => Unit
  ): Unit = {
    var excluded = Set.empty[Vertex]
    for (v <- degeneracy_ordering[Vertex](graph, -1)) {
      val neighbours = graph.neighbours(v)
      assert(neighbours.nonEmpty)
      val neighbouring_candidates = neighbours &~ excluded
      if (neighbouring_candidates.nonEmpty) {
        val neighbouring_excluded = util.intersect(neighbours, excluded)
        visit(
          graph,
          reporter,
          Arbitrary,
          Arbitrary,
          neighbouring_candidates,
          neighbouring_excluded,
          immutable.List(v)
        )
      } else {
        assert(!util.are_disjoint(neighbours, excluded))
      }
      excluded += v
    }
  }
}
